contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,313 @@
+        - [Set Up Your Editor for Commit Messages](#set-up-your-editor-for-commit-messages)
+    - [Always Commit Functional Code](#always-commit-functional-code)
+        - [Stashing](#stashing)
+    - [One Logical Change to One Commit](#one-logical-change-to-one-commit)
+    - [Hide the Sausage Making](#hide-the-sausage-making)
+        - [How to Hide the Sausage](#how-to-hide-the-sausage)
+    - [Respect Published History](#respect-published-history)
+    - [Keep Up To Date](#keep-up-to-date)
+        - [Working with Remotes](#working-with-remotes)
+        - [Rebase versus Merging](#rebase-versus-merging)
+- [Resources](#resources)
+    - [Further Reading](#further-reading)
+
+
+# Introduction
+[`git`](https://git-scm.com/) is a popular source code management tool and commonly used in many open-source projects, especially those on [GitHub](https://github.com).
+
+# Best Practices
+`git` can be an incredibly useful tool for collaboration or it can be a terrible headache. Best practices exist in order to create a common understanding between users so that the latter does not happen.
+
+Best practices are guidelines that are mostly sensible, but are still guidelines. You can always choose to ignore them although best if you have a compelling reason to do so.
+
+---
+
+## Write Good Commit Messages
+Good commit messages can help reviewers or other contributors to understand:
+- the high-level changes made by your pull request/patch
+- the reasoning behind the changes made by that commit while
+    - reviewing your code
+    - figuring out why a piece of code that is five years old is that way
+
+They also assist you in the development process if you forget what has been done, or if you need to cherry-pick commits for elsewhere.
+
+### What Constitutes a Good Commit Message
+The easiest way to attain commit message discipline is to stop putting in one-liner descriptions using `git commit -m "Add some things to that."`. Instead, write a proper commit message in an editor:
+
+```bash
+# opens your editor to write a commit message properly
+git add files-to-stage
+git commit
+
+# like above, but shows the diff of the currently staged files
+git add files-to-stage
+git commit --verbose
+
+# amend the most recent commit message
+git commit --amend HEAD^
+```
+
+A good commit message can be formatted the following way:
+```
+Capitalized, short (50 chars or less) summary
+
+More detailed explanatory text, if necessary.  Wrap it to about 72
+characters or so.  In some contexts, the first line is treated as the
+subject of an email and the rest of the text as the body.  The blank
+line separating the summary from the body is critical (unless you omit
+the body entirely); tools like rebase can get confused if you run the
+two together.
+
+Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
+or "Fixes bug."  This convention matches up with commit messages generated
+by commands like git merge and git revert.
```

Comment:
This may be useful https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

Reply:
I decided to scrap the old format for describing a good commit message and I took the list from that link and reorganized it a bit.